### PR TITLE
Utilies to simplify async command use case

### PR DESCRIPTION
**Pull request by slinkydeveloper**
This PR contains 2 new features to simplify the creation of async commands.

You can look at a complete example: https://github.com/knative-sandbox/control-protocol/pull/49/files#diff-9603966e6fea4c98a699272300e352d9c9ce7cded757201ec91848583677edd7R121

Check the docs: https://github.com/knative-sandbox/control-protocol/blob/e1221c1f1f08c094a2b3982b61e160518da3684e/docs/async_command_pattern.md

Part of #47

# Changes

- :gift: Add `NewAsyncCommandHandler` to create commands that acks immediately, but notifies asynchronously of their result
- :gift: Add `AsyncCommandNotificationStore` as a wrapper of `NotificationStore`, specialized to handle `AsyncCommandResult` notifications.
- :gift: Draft of the docs for this use case

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
/hold

I wanna test it in https://github.com/knative-sandbox/eventing-kafka/pull/530 first

File: docs/async_command_pattern.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Async Command pattern
+
+The async command pattern allows developing commands where the controller sends a command to the data plane and doesn't expect an immediate response.
+
+This is similar to the request-response, with the important difference that the command execution doesn't block the reconciliation, that is the command may take a while to execute, and in the meantime the reconciler can flag the k8s resource as _processing_ such command.
+
+The flow of the async command pattern looks like:
+
+<!-- PlantUML source:
```
Comment (matzew) [inline docs/async_command_pattern.md:9]: UML :eyes:

File: docs/async_command_pattern.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+@enduml
+
+-->
+
+![](https://plantuml-server.kkeisuke.dev/svg/ZP7BRW8n34Nt_WgBBKBTpmA1gBgkoXTOZcV6QYP6YOFKls-01GEQFdPUnDVtdEoAK_OwHG1YrpEvuC6IPujHCjn7t6nnzKfEU8gKP8NhTOT7IG7tvIlnmQQ9KW1uUDDsxWaTxlaJahKBKNhly2tIW3uAVaYncbcG2fwoiPIYQO0WIvMk0NPkZURHnz6oRrWpLtNCmfPOevAh9RZjP1r6H-iVC3fylnsy5k6_APQv6q6D3h_u-XzzgSmI9Bpqf572NC4y37wmS9arLNaMi6mITWsZVVqt.svg)
+
+In order to implement this pattern, some utilities are provided:
+
+* `service.NewAsyncCommandHandler` which wraps `ServiceMessage` in order to simplify sending back the async result:
```
Comment (matzew) [inline docs/async_command_pattern.md:40]: `ServiceMessage` ... not 100% sold on that name :smile_cat:  but that's more details :-)

Comment (slinkydeveloper) [reply docs/async_command_pattern.md:40]: `ServiceMessage` is comes from `Service` interface, to distinguish from `Message`... If you don't like it, can you open an issue and propose a better name? :)

Comment (matzew) [reply docs/async_command_pattern.md:40]: yeah, I saw the service in there later on... all good.

(service still trigger an uncomfortable feeling in me :smile: )

File: pkg/service/async_command_handler.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+func NewAsyncCommandHandler(svc control.Service, asyncCommandType message.AsyncCommand, resultOpCode control.OpCode, handler func(context.Context, AsyncCommandMessage)) control.MessageHandler {
+	return &asyncCommandHandler{
+		svc:              svc,
+		resultOpCode:     resultOpCode,
+		asyncCommandType: reflect.Indirect(reflect.ValueOf(asyncCommandType)).Type(),
+		handler:          handler,
+	}
+}
+
+func (f *asyncCommandHandler) HandleServiceMessage(ctx context.Context, msg control.ServiceMessage) {
```
Comment (matzew) [inline pkg/service/async_command_handler.go:104]: wanna add some comments on this func?

File: pkg/service/async_command_handler.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	// This is a pointer to asyncCommandType
+	commandValue := reflect.New(f.asyncCommandType).Interface()
+
+	// This assertion is always true b/c asyncCommandType is per contract a message.AsyncCommand,
+	// which implements encoding.BinaryUnmarshaler
+	err := commandValue.(encoding.BinaryUnmarshaler).UnmarshalBinary(msg.Payload())
```
Comment (matzew) [inline pkg/service/async_command_handler.go:110]: wanna name `commandValue` better as `asyncCommandType` than? (b/c of the comments)

File: pkg/test/service_mock.go [hunk 5]
```diff
@@ -73,6 +73,17 @@ func (s *ServiceMock) InvokeMessageHandler(ctx context.Context, message *control
...
 
+// InvokeMessageHandlerWithErr invokes the registered message handler and returns the eventual error if the message was acked back with an error
+func (s *ServiceMock) InvokeMessageHandlerWithErr(ctx context.Context, message *control.Message) error {
+	acked := atomic.NewError(nil)
+	ackFn := func(err error) {
+		acked.Store(err)
+	}
+
+	s.messageHandler.HandleServiceMessage(ctx, control.NewServiceMessage(message, ackFn))
+	return acked.Load()
```
Comment (matzew) [inline pkg/test/service_mock.go:84]: load... LOL ... I would have expected `get()` on this atomic boolean

File: pkg/service/async_command_handler.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		logging.FromContext(ctx).Warnw("Error while parsing the async command", zap.Error(err))
+		msg.AckWithError(fmt.Errorf("error while parsing the async command: %w", err))
+		return
+	}
+
+	asyncCommandMsg := AsyncCommandMessage{
+		serviceMessage: msg,
+		parsedCommand:  commandValue.(message.AsyncCommand),
```
Comment (matzew) [inline pkg/service/async_command_handler.go:119]: why `parsedCommand` ?  isn't that just the command (or command code), of the message?

**Comment by travis-minke-sap** [comment]
👍  looking good!

File: docs/async_command_pattern.md (hunk 2 above)
Comment (slinkydeveloper) [reply docs/async_command_pattern.md:40]: @matzew I really suggest you to propose another name then in an issue, those names have to be comfortable for people :smile:

File: pkg/service/async_command_handler.go (hunk 4 above)
Comment (slinkydeveloper) [reply pkg/service/async_command_handler.go:110]: this one is a value, not a type, let me fix the comment instead

File: pkg/service/async_command_handler.go (hunk 6 above)
Comment (slinkydeveloper) [reply pkg/service/async_command_handler.go]: nope, renamed the vars to make it clear

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Pull request merged**